parsers/rhea/rhea.go

```diff
@@ -0,0 +1,791 @@
+type EC struct {
+	XMLName  xml.Name `xml:"ec"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// Status is an XML representation of the current status of an description.
+type Status struct {
+	XMLName  xml.Name `xml:"status"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// CompoundXML is an XML representation of a compound.
+type CompoundXML struct {
+	XMLName  xml.Name `xml:"compound"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// BidirectionalReaction is an XML representation of a Rhea bidirectional reaction.
+type BidirectionalReaction struct {
+	XMLName  xml.Name `xml:"bidirectionalReaction"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// DirectionalReaction is an XML representation of a Rhea directional reaction.
+type DirectionalReaction struct {
+	XMLName  xml.Name `xml:"directionalReaction"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// Side is an XML representation of a Rhea ReactionSide.
+type Side struct {
+	XMLName  xml.Name `xml:"side"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// SeeAlso is an XML representation of a SeeAlso XML in a description.
+type SeeAlso struct {
+	XMLName  xml.Name `xml:"seeAlso"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// TransformableTo is an XML representation of a transformableTo in a description. This essentially links two ReactionSides in Rhea.
+type TransformableTo struct {
+	XMLName  xml.Name `xml:"transformableTo"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// Contains is an XML representation of what Compound a ReactionParticipant contains.
+type Contains struct {
+	XMLName  xml.Name `xml:"contains"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// ContainsX is a catch-all XML representation of how many compounds a ReactionParticipant would use.
+type ContainsX struct {
+	XMLName xml.Name
+	Content string `xml:"resource,attr"`
+}
+
+// ChebiXML is an XML representation of a Chebi.
+type ChebiXML struct {
+	XMLName  xml.Name `xml:"chebi"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// UnderlyingChebi is an XML representation of Chebi that builds a Polymer.
+type UnderlyingChebi struct {
+	XMLName  xml.Name `xml:"underlyingChebi"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// ReactivePartXML is an XML representation of a ReactivePart.
+type ReactivePartXML struct {
+	XMLName  xml.Name `xml:"reactivePart"`
+	Resource string   `xml:"resource,attr"`
+}
+
+// Location is an XML representation of Locations in a cell, usually referring to transport enzymes.
+type Location struct {
+	XMLName  xml.Name `xml:"location"`
+	Resource string   `xml:"resource,attr"`
+}
+
+/******************************************************************************
+
+Higher level structs
+
+These structs are what you would put into a database or directly use. In order to
+create a tree or insert into a normalized database, you would insert in the following
+order:
+
+	- Compound
+	- ReactionSide
+	- ReactionParticipant
+	- Reaction
+	- ReactionSide <-> Reaction
```

Comment:
Thinking about this some more.

Each of these higher level struct should have a `constructor` function that should be the only way a new one should be created.

```
func NewCompound( ... ) *Compound { ... }
func NewReactionSide( ... ) *ReactionSide { ... }
...
```
Each of these should be used in your large parsing switch statement - the refactoring we where discussing here: https://github.com/TimothyStiles/poly/pull/108/files#r633038151

Each of these `Constructors` should be unit tested and validate all of the constraint you discovered and are representing with your SQL tables. e.g. constraints like: https://github.com/TimothyStiles/poly/pull/108/files#diff-3ffc125997632ea24ce7c093b059cbb438eec4249c802bc3519f21063c1343d8R560

Enforcing these constraints in code instead of a one time SQL check will be make things more readable and idiomatic. It will also enforce correct behavior on users who may create these thing dynamically and not wish to regularly commit their results to SQL for a validation check.

Reply:
Ok, going to do that after this next push.